File: tools/gen_modelbin_web/frontend/js/tabs/glyph-template.js
// Glyph Template tab: generate a KFPS-importable glyph template for a
// font. Second consumer of the shared font-search component.
window.ForzaTabs = window.ForzaTabs || {};

(function () {
  function esc(s) {
    return String(s).replace(/[&<>"']/g, (c) => (
      { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[c]
    ));
  }

  function hexToRgb(hex) {
    let h = hex.trim().replace('#', '');
    if (h.length === 3) h = h.split('').map((c) => c + c).join('');
    if (!/^[0-9a-f]{6}$/i.test(h)) return null;
    const n = parseInt(h, 16);
    return [(n >> 16) & 255, (n >> 8) & 255, n & 255];
  }
  function rgbaToHex(rgba) {
    return '#' + [rgba[0], rgba[1], rgba[2]].map((c) => c.toString(16).padStart(2, '0')).join('');
  }

  const CHARSET_LABELS = { 'basic-latin': 'Basic Latin', 'hiragana': 'Hiragana', 'katakana': 'Katakana' };
  function charsetLabel(key) {
    return CHARSET_LABELS[key] || key;
  }

  async function mount(container) {
    container.innerHTML = `
      <h2 class="page-heading">Glyph Template</h2>
      <div class="intro-text">
        Generate a labeled uniform grid for your selected font, in the form of an SVG Template.<br>
        This can be used in the KFPS Fabric Editor, or used as a tracing guide in the game.
      </div>

      <div class="gi-content" style="grid-template-columns: 1fr 340px;">
        <div>
      <div class="section">
        <div class="section-title">1. Font</div>
        <div style="display:flex; gap: 10px; align-items: center; flex-wrap: wrap; margin-bottom: 8px;">
          <div id="gtFontSearch"></div>
          <button type="button" class="btn" id="gtBrowseFont">Browse file…</button>
        </div>
        <div class="field-hint" id="gtFontStatus" style="margin-bottom: 10px;">Select a font to build a template for.</div>
        <div class="field-hint" style="color: var(--warn);">
          The font file is embedded directly into the generated SVG/project as a tracing guide.
          Only use a font you actually hold a license for.
        </div>
      </div>

      <div class="section">
        <div class="section-title">2. Charset</div>
        <div class="field-row" style="margin-bottom: 6px;">
          <label><input type="radio" name="gtMode" value="single" checked> Single template</label>
          <label style="margin-left: 14px;">Chars/row <input type="number" class="path-input" id="gtCharsPerRow" style="width: 60px; margin-left: 6px;"></label>
        </div>
        <div id="gtCharsetGroup" style="display:flex; flex-direction:column; gap:8px; margin: 0 0 8px 24px;"></div>
        <div class="field-row" style="margin-bottom: 6px;">
          <label><input type="radio" name="gtMode" value="split"> Split by Unicode block (recommended for large libraries)</label>
          <label style="margin-left: 14px;">Min glyphs/block <input type="number" class="path-input" id="gtMinChars" style="width: 60px; margin-left: 6px;"></label>
        </div>
        <div class="field-hint" style="margin-bottom: 8px;">
          Split writes one template per Unicode block the font actually covers, its own folder and
          SVG each, so a large or non-Latin font does not produce one unwieldy grid.
        </div>
        <div id="gtBlocksWrap" style="display:none;">
          <div id="gtBlocksList" style="max-height: 220px; overflow-y: auto; margin-bottom: 6px;"></div>
        </div>
        <div class="field-hint" id="gtBlocksStatus">Load a font to see which Unicode blocks it covers.</div>
      </div>

      <div class="section">
        <div class="section-title">3. Output</div>
        <div class="field-row">
          <div class="field-label">Prefix</div>
          <input type="text" class="path-input" id="gtPrefix" value="CUSTOM" style="max-width: 300px;">
        </div>
        <div class="field-row">
          <div class="field-label">Output folder</div>
          <div class="path-field">
            <input type="text" class="path-input" id="gtOutDir">
            <button type="button" class="btn" id="gtBrowseOutDir">Browse…</button>
          </div>
        </div>
        <ul class="field-hint field-hint-list">
          <li>A separate folder from Fontpacks Output Directory. These are blank tracing templates, not finished fontpacks.</li>
          <li>Single mode writes &lt;folder&gt;/&lt;prefix&gt;/.</li>
          <li>Split mode writes one &lt;folder&gt;/&lt;prefix&gt;/&lt;prefix&gt;-&lt;BLOCK&gt;/ per checked block.</li>
        </ul>
      </div>

      <div class="field-row" style="flex-direction: row; align-items: center; gap: 10px;">
        <button type="button" class="btn accent" id="gtGenerate">Generate</button>
        <span class="field-hint" id="gtStatus"></span>
      </div>
        </div>

        <div>
          <div class="section">
            <div class="section-title">Preview</div>
            <div class="gt-preview-panel" id="gtPreviewPanel">
              <div class="gt-preview-empty" id="gtPreviewEmpty">Load a font to see a preview.</div>
              <img id="gtPreviewImg" style="display:none;">
            </div>
            <div class="field-hint" id="gtPreviewHint" style="margin-top: 6px;"></div>
          </div>

          <div class="section">
            <div class="section-title">Tracing Color</div>
            <div style="display:flex; align-items: center; gap: 10px; margin-bottom: 8px;">
              <div class="cp-swatch" id="gtColorSwatch" style="width: 36px; height: 24px; cursor: pointer;"></div>
              <input type="text" class="path-input" id="gtColorHex" style="width: 90px;">
              <input type="color" id="gtColorNative" style="display:none;">
            </div>
            <div class="field-hint" style="margin-bottom: 10px;">
              Fill color of the traced letterform in the generated SVG. Type a #rgb/#rrggbb hex code,
              or pick one below. Choose something that stands out while tracing over it in KFPS.
            </div>
            <div id="gtColorPicker"></div>
          </div>
        </div>
      </div>
    `;

    const api = window.pywebview.api;
    let loadedFont = null; // {path, name, summary, covered, suggested_mode, suggested_prefix}
    let prefixEdited = false;
    let pendingGeneration = null;

    // -- charset / defaults -------------------------------------------------
    // Card style matches Direct Generator's method cards (direct.js's
    // renderMethodCards): a plain clickable div per option, no visible radio
    // dot, selected state shown purely via the orange border/title color.
    let selectedCharset = null;
    function currentCharset() { return selectedCharset || 'basic-latin'; }
    function renderCharsetCards(charsets) {
      container.querySelector('#gtCharsetGroup').innerHTML = charsets.map((c) => `
        <div class="section method-card ${currentCharset() === c ? 'active' : ''}" data-key="${esc(c)}"
             style="margin-bottom:0; cursor:pointer; padding: 8px 14px; ${currentCharset() === c ? 'border-color: var(--accent);' : ''}">
          <div style="font-family: var(--display); font-weight: 600; font-size: 12px; ${currentCharset() === c ? 'color: var(--accent);' : ''}">${esc(charsetLabel(c))}</div>
        </div>
      `).join('');
      container.querySelector('#gtCharsetGroup').querySelectorAll('.method-card').forEach((card) => {
        card.addEventListener('click', () => {
          selectedCharset = card.dataset.key;
          renderCharsetCards(charsets);
          schedulePreviewRefresh();
        });
      });
    }

    const charsetsResp = await api.call('glyph_template.get_charsets', {});
    let defaultMinChars = 4;
    if (charsetsResp.ok) {
      const { charsets, default_chars_per_row, default_trace_color, default_min_chars, default_output_dir } = charsetsResp.result;
      selectedCharset = charsets[0] || 'basic-latin';
      renderCharsetCards(charsets);
      container.querySelector('#gtCharsPerRow').value = default_chars_per_row;
      container.querySelector('#gtMinChars').value = default_min_chars;
      defaultMinChars = default_min_chars;
      container.querySelector('#gtColorHex').value = default_trace_color;
      container.querySelector('#gtColorSwatch').style.background = default_trace_color;
      // A separate default from Settings' Fontpacks Output Directory --
      // these are blank tracing templates, not finished fontpacks.
      container.querySelector('#gtOutDir').value = default_output_dir;
    }

    // -- tracing color --------------------------------------------------------
    // The full shared picker (saved/recent colors, HSB sliders, OS picker)
    // external-drives off the same #gtColorHex value the compact swatch and
    // the generate payload both already read from -- one source of truth,
    // not a second color state to keep in sync. colorPicker is declared
    // before setColor can be reached by any event so a click racing the
    // picker's own (awaited) setup never hits it uninitialized.
    let colorPicker = null;
    function setColor(hex) {
      container.querySelector('#gtColorHex').value = hex;
      container.querySelector('#gtColorSwatch').style.background = hex;
      if (colorPicker) colorPicker.sync();
      schedulePreviewRefresh();
    }
    container.querySelector('#gtColorSwatch').addEventListener('click', () => {
      container.querySelector('#gtColorNative').value = container.querySelector('#gtColorHex').value || '#e6e6e6';
      container.querySelector('#gtColorNative').click();
    });
    container.querySelector('#gtColorNative').addEventListener('change', (e) => setColor(e.target.value));
    container.querySelector('#gtColorHex').addEventListener('change', (e) => {
      if (/^#([0-9a-f]{3}|[0-9a-f]{6})$/i.test(e.target.value.trim())) setColor(e.target.value.trim());
    });

    colorPicker = await window.ForzaColorPicker.create(container.querySelector('#gtColorPicker'), {
      title: '',
      getColor: () => {
        const hex = container.querySelector('#gtColorHex').value.trim();
        return [...(hexToRgb(hex) || [230, 230, 230]), 255];
      },
      onChange: (rgba) => setColor(rgbaToHex(rgba)),
    });

    // -- font loading -----------------------------------------------------
    function applyLoadedFont(font) {
      loadedFont = font;
      container.querySelector('#gtFontStatus').textContent = font.summary;
      if (!prefixEdited) container.querySelector('#gtPrefix').value = font.suggested_prefix;
      container.querySelector(`input[name="gtMode"][value="${font.suggested_mode}"]`).checked = true;
      updateModeVisibility();
      renderBlockChecklist();
    }

    window.ForzaFontSearch.create(container.querySelector('#gtFontSearch'), {
      placeholder: 'Search installed fonts…',
      onSelect: async (font) => {
        container.querySelector('#gtFontStatus').textContent = `Loading ${font.name}…`;
        const resp = await api.call('glyph_template.load_font_by_path', { path: font.path });
        if (!resp.ok) { container.querySelector('#gtFontStatus').textContent = resp.error; return; }
        applyLoadedFont(resp.result);
      },
    });
    container.querySelector('#gtBrowseFont').addEventListener('click', async () => {
      container.querySelector('#gtFontStatus').textContent = 'Loading font…';
      const resp = await api.call('glyph_template.browse_font', {});
      if (!resp.ok) { container.querySelector('#gtFontStatus').textContent = resp.error; return; }
      if (resp.result.cancelled) return;
      applyLoadedFont(resp.result);
    });

    container.querySelector('#gtPrefix').addEventListener('input', () => { prefixEdited = true; });

    // -- live grid preview --------------------------------------------------
    // Split mode writes one grid per checked block, so there is no single
    // "the" grid to show without one chosen -- clicking a block's row below
    // (not its checkbox, which stays purely include/exclude) sets it as the
    // one currently previewed. Single mode has no such ambiguity: the
    // preview always reflects the one selected charset.
    let previewBlockName = null;
    let previewDebounceId = null;
    let previewRequestId = 0;

    function schedulePreviewRefresh() {
      if (previewDebounceId) clearTimeout(previewDebounceId);
      previewDebounceId = setTimeout(refreshPreview, 180);
    }

    function showPreviewEmpty(message) {
      container.querySelector('#gtPreviewImg').style.display = 'none';
      const emptyEl = container.querySelector('#gtPreviewEmpty');
      emptyEl.textContent = message;
      emptyEl.style.display = '';
      container.querySelector('#gtPreviewHint').textContent = '';
    }

    async function refreshPreview() {
      if (!loadedFont) { showPreviewEmpty('Load a font to see a preview.'); return; }
      const mode = container.querySelector('input[name="gtMode"]:checked').value;
      const payload = {
        font_path: loadedFont.path,
        text_color: container.querySelector('#gtColorHex').value.trim() || '#e6e6e6',
        chars_per_row: Number(container.querySelector('#gtCharsPerRow').value || 10),
        mode,
      };
      if (mode === 'split') {
        const block = loadedFont.covered.find((b) => b.name === previewBlockName);
        if (!block) { showPreviewEmpty('Click a block below to preview its grid.'); return; }
        payload.block = { name: block.name, chars: block.chars };
      } else {
        payload.charset = currentCharset();
      }

      const requestId = ++previewRequestId;
      const resp = await api.call('glyph_template.render_preview', payload);
      if (requestId !== previewRequestId) return; // superseded by a newer request
      if (!resp.ok || !resp.result.image) {
        showPreviewEmpty(resp.ok ? 'No preview available.' : resp.error);
        return;
      }
      container.querySelector('#gtPreviewImg').src = resp.result.image;
      container.querySelector('#gtPreviewImg').style.display = '';
      container.querySelector('#gtPreviewEmpty').style.display = 'none';
      container.querySelector('#gtPreviewHint').textContent = mode === 'split'
        ? `Previewing ${previewBlockName} — click another block below to switch.` : '';
    }

    // -- mode / block checklist --------------------------------------------
    function updateModeVisibility() {
      const mode = container.querySelector('input[name="gtMode"]:checked').value;
      container.querySelector('#gtBlocksWrap').style.display = mode === 'split' ? '' : 'none';
    }
    container.querySelectorAll('input[name="gtMode"]').forEach((r) => r.addEventListener('change', () => {
      updateModeVisibility();
      schedulePreviewRefresh();
    }));

    function setPreviewBlock(name) {
      previewBlockName = name;
      container.querySelectorAll('.gt-block-row').forEach((row) => {
        row.classList.toggle('gt-block-previewing', row.dataset.blockName === name);
      });
      schedulePreviewRefresh();
    }

    async function refreshBlockSamples(blocks) {
      if (!loadedFont || blocks.length === 0) return;
      const resp = await api.call('glyph_template.render_block_samples', {
        font_path: loadedFont.path,
        blocks: blocks.map((b) => ({ name: b.name, chars: b.chars })),
      });
      if (!resp.ok) return;
      container.querySelectorAll('.gt-block-row').forEach((row) => {
        const uri = resp.result.samples[row.dataset.blockName];
        if (uri) row.querySelector('.gt-block-sample').style.backgroundImage = `url(${uri})`;
      });
    }

    function renderBlockChecklist() {
      const listEl = container.querySelector('#gtBlocksList');
      const statusEl = container.querySelector('#gtBlocksStatus');
      if (!loadedFont) {
        statusEl.textContent = 'Load a font to see which Unicode blocks it covers.';
        listEl.innerHTML = '';
        previewBlockName = null;
        return;
      }
      const minChars = Math.max(1, Number(container.querySelector('#gtMinChars').value || defaultMinChars));
      const eligible = loadedFont.covered.filter((b) => b.chars.length >= minChars);
      if (eligible.length === 0) {
        statusEl.textContent = `No block reaches ${minChars} glyph(s) at this threshold. Lower "Min glyphs/block".`;
        listEl.innerHTML = '';
        previewBlockName = null;
        schedulePreviewRefresh();
        return;
      }
      listEl.innerHTML = eligible.map((b) => `
        <div class="checkbox-row gt-block-row" data-block-name="${esc(b.name)}">
          <input type="checkbox" checked data-block="${esc(b.name)}" id="gtBlock_${esc(b.name)}">
          <div class="gt-block-sample"></div>
          <label for="gtBlock_${esc(b.name)}">${esc(b.name)} (${b.chars.length})</label>
        </div>
      `).join('');
      statusEl.textContent =
        `${eligible.length} of ${loadedFont.total_known_blocks} known block(s) covered at this threshold. ` +
        'Uncheck any you do not want a template for.';

      listEl.querySelectorAll('.gt-block-row').forEach((row) => {
        row.addEventListener('click', () => setPreviewBlock(row.dataset.blockName));
      });
      if (!previewBlockName || !eligible.some((b) => b.name === previewBlockName)) {
        previewBlockName = eligible[0].name;
      }
      listEl.querySelectorAll('.gt-block-row').forEach((row) => {
        row.classList.toggle('gt-block-previewing', row.dataset.blockName === previewBlockName);
      });
      refreshBlockSamples(eligible);
      schedulePreviewRefresh();
    }
    container.querySelector('#gtMinChars').addEventListener('input', renderBlockChecklist);
    container.querySelector('#gtCharsPerRow').addEventListener('input', schedulePreviewRefresh);

    container.querySelector('#gtBrowseOutDir').addEventListener('click', async () => {
      const resp = await api.call('glyph_template.pick_output_dir', { initial: container.querySelector('#gtOutDir').value });
      if (resp.ok && !resp.result.cancelled) container.querySelector('#gtOutDir').value = resp.result.path;
    });

    // -- generate -----------------------------------------------------------
    container.querySelector('#gtGenerate').addEventListener('click', async () => {
      const statusEl = container.querySelector('#gtStatus');
      if (!loadedFont) { statusEl.textContent = 'Select a font first.'; return; }
      const textColor = container.querySelector('#gtColorHex').value.trim();
      if (!/^#([0-9a-f]{3}|[0-9a-f]{6})$/i.test(textColor)) {
        statusEl.textContent = `${textColor} is not a valid #rgb or #rrggbb hex color.`;
        return;
      }
      const mode = container.querySelector('input[name="gtMode"]:checked').value;
      const payload = {
        font_path: loadedFont.path,
        text_color: textColor,
        prefix: container.querySelector('#gtPrefix').value,
        out_dir: container.querySelector('#gtOutDir').value,
        mode,
        chars_per_row: container.querySelector('#gtCharsPerRow').value,
      };
      if (mode === 'split') {
        payload.min_chars = container.querySelector('#gtMinChars').value;
        payload.only_blocks = Array.from(container.querySelectorAll('#gtBlocksList input:checked')).map((el) => el.dataset.block);
        if (payload.only_blocks.length === 0) { statusEl.textContent = 'Check at least one Unicode block to generate.'; return; }
      } else {
        payload.charset = currentCharset();
      }

      container.querySelector('#gtGenerate').disabled = true;
      statusEl.textContent = 'Generating…';
      const resp = await api.call('glyph_template.generate', payload);
      if (!resp.ok) {
        statusEl.textContent = resp.error;
        container.querySelector('#gtGenerate').disabled = false;
        return;
      }
      pendingGeneration = resp.result.generation;
    });

    const offLog = window.__forzaEvents.on('glyph_template_log', (generation, payload) => {
      if (generation !== pendingGeneration) return;
      container.querySelector('#gtStatus').textContent = payload.line;
    });
    const offDone = window.__forzaEvents.on('glyph_template_done', (generation, payload) => {
      if (generation !== pendingGeneration) return;
      container.querySelector('#gtStatus').textContent = payload.message;
      container.querySelector('#gtGenerate').disabled = false;
    });

    return () => { offLog(); offDone(); colorPicker.destroy(); };
  }

  window.ForzaTabs.glyph_template = mount;
})();

File: tools/gen_modelbin_web/handlers/glyph_template.py
"""Glyph Template tab: generate a KFPS-importable glyph template for a
font. A thin wrapper around the same two CLI tools the Tkinter tab already
wraps (tools/gen_glyph_template.py, tools/gen_font_block_templates.py) --
nothing here reimplements grid/SVG/font-embedding logic.
"""
from __future__ import annotations

import sys
import threading
from pathlib import Path

import webview

_TOOLS_DIR = Path(__file__).resolve().parent.parent.parent
_REPO_ROOT = _TOOLS_DIR.parent
sys.path.insert(0, str(_TOOLS_DIR))
sys.path.insert(0, str(_REPO_ROOT))

import gen_font_block_templates  # noqa: E402
import gen_glyph_template  # noqa: E402
import gui_settings  # noqa: E402
from gen_fontpack import sanitize_prefix  # noqa: E402
from fontTools.ttLib import TTFont  # noqa: E402
from forza_writer.fabric_project import save as save_project  # noqa: E402
from forza_writer.font_info import load_font_info  # noqa: E402
from forza_writer.glyph_template import (  # noqa: E402
    DEFAULT_CHARS_PER_ROW, DEFAULT_TRACE_TEXT_COLOR, TEMPLATE_UNICODE_BLOCKS, blocks_covered_by_font,
    save_template, validate_hex_color)

from ..events import push_event  # noqa: E402

_FONT_FILE_TYPES = ('Fonts (*.ttf;*.otf;*.ttc)',)
# The first four TEMPLATE_UNICODE_BLOCKS entries are exactly "Basic Latin"
# split into uppercase/lowercase/digits/punctuation -- see
# tools/gen_modelbin_gui/tabs/glyph_template.py's own comment on this.
_BASIC_LATIN_BLOCK_NAMES = frozenset(name for name, _ranges in TEMPLATE_UNICODE_BLOCKS[:4])


def _load_and_scan(font_path: Path) -> dict:
    info = load_font_info(font_path)
    font = TTFont(str(font_path), fontNumber=0)
    try:
        cmap = font.getBestCmap() or {}
    finally:
        font.close()
    # min_chars=1: the full inventory of covered blocks. The UI's own
    # min-glyphs-per-block threshold filters this client-side as the user
    # changes it, without re-scanning the font each time.
    covered = blocks_covered_by_font(cmap, min_chars=1)
    glyph_count = sum(len(chars) for _name, chars in covered)
    weight = f', weight {info.names.weight_class}' if info.names.weight_class else ''
    italic = ', italic' if info.names.is_italic else ''
    beyond_basic_latin = any(name not in _BASIC_LATIN_BLOCK_NAMES for name, _chars in covered)
    return {
        'path': str(font_path),
        'name': info.names.full_name,
        'summary': (f'{info.names.full_name} ({info.names.family} {info.names.subfamily}'
                    f'{weight}{italic}). {glyph_count:,} glyph(s) across {len(covered)} '
                    f'Unicode block(s).'),
        'covered': [{'name': name, 'chars': chars} for name, chars in covered],
        'suggested_mode': 'split' if beyond_basic_latin else 'single',
        'suggested_prefix': sanitize_prefix(info.names.family),
        'total_known_blocks': len(TEMPLATE_UNICODE_BLOCKS),
    }


def register(api, window) -> None:
    generation = {'n': 0}

    def get_charsets(_payload: dict) -> dict:
        return {
            'charsets': list(gen_glyph_template.CHARSETS),
            'default_chars_per_row': DEFAULT_CHARS_PER_ROW,
            'default_trace_color': DEFAULT_TRACE_TEXT_COLOR,
            'default_min_chars': gen_font_block_templates.DEFAULT_MIN_CHARS,
            # A separate default from Settings' Fontpacks Output Directory --
            # these are blank tracing templates, not finished fontpacks.
            'default_output_dir': gui_settings.load_settings()['glyph_template_output_dir'],
        }

    def browse_font(_payload: dict) -> dict:
        chosen = window.create_file_dialog(webview.FileDialog.OPEN, file_types=_FONT_FILE_TYPES)
        if not chosen:
            return {'cancelled': True}
        return _load_and_scan(Path(chosen[0]))

    def load_font_by_path(payload: dict) -> dict:
        return _load_and_scan(Path(payload['path']))

    def pick_output_dir(payload: dict) -> dict:
        chosen = window.create_file_dialog(
            webview.FileDialog.FOLDER, directory=payload.get('initial', ''))
        if not chosen:
            return {'cancelled': True}
        return {'path': chosen[0]}

    def generate(payload: dict) -> dict:
        font_path = Path(payload['font_path'])
        text_color = payload['text_color']
        validate_hex_color(text_color)  # raises ValueError -> surfaced to the page as an error
        prefix = sanitize_prefix(payload['prefix'])
        out_dir = Path(payload['out_dir'])
        gui_settings.update_settings({'glyph_template_output_dir': str(out_dir)})
        mode = payload['mode']
        chars_per_row = max(1, int(payload['chars_per_row']))

        generation['n'] += 1
        gen = generation['n']

        def post_log(line: str) -> None:
            push_event(window, 'glyph_template_log', gen, {'line': line})

        def worker():
            try:
                if mode == 'split':
                    only_blocks = set(payload['only_blocks'])
                    if not only_blocks:
                        raise ValueError('Check at least one Unicode block to generate.')
                    min_chars = max(1, int(payload['min_chars']))
                    written = []
                    for _block_name, template_id, template, project in (
                            gen_font_block_templates.build_all_block_projects(
                                font_path, prefix, chars_per_row, min_chars, only_blocks,
                                log=post_log, text_color=text_color)):
                        # Nested under out_dir/prefix/ -- not out_dir/template_id/ directly --
                        # so a font's blocks group under one folder instead of scattering as
                        # loose siblings alongside every other font's own block folders.
                        block_dir = out_dir / prefix / template_id
                        save_template(template, block_dir / f'{template_id}_template.json')
                        save_project(project, block_dir / f'{template_id}_blank.fabric-project.json')
                        svg_path = block_dir / f'{template_id}.svg'
                        svg_path.parent.mkdir(parents=True, exist_ok=True)
                        svg_path.write_text(project['editor_source_overlay']['svg_text'], encoding='utf-8')
                        written.append(template_id)
                    message = f'{len(written)} block template(s) written under {out_dir / prefix}.'
                else:
                    charset = payload['charset']
                    template, project = gen_glyph_template.build_blank_project(
                        prefix, chars_per_row, charset, font_path, log=post_log, text_color=text_color)
                    pack_dir = out_dir / prefix
                    save_template(template, pack_dir / f'{prefix}_template.json')
                    save_project(project, pack_dir / f'{prefix}_blank.fabric-project.json')
                    svg_path = pack_dir / f'{prefix}.svg'
                    svg_path.parent.mkdir(parents=True, exist_ok=True)
                    svg_path.write_text(project['editor_source_overlay']['svg_text'], encoding='utf-8')
                    message = (f'{len(template.slots)} slot(s) across {template.row_count} row(s) '
                               f'-> {pack_dir}.')
                push_event(window, 'glyph_template_done', gen, {'ok': True, 'message': message})
            except Exception as exc:
                push_event(window, 'glyph_template_done', gen, {'ok': False, 'message': str(exc)})

        threading.Thread(target=worker, daemon=True).start()
        return {'generation': gen}

    api.register('glyph_template.get_charsets', get_charsets)
    api.register('glyph_template.browse_font', browse_font)
    api.register('glyph_template.load_font_by_path', load_font_by_path)
    api.register('glyph_template.pick_output_dir', pick_output_dir)
    api.register('glyph_template.generate', generate)
